### PR TITLE
Make error types unexported

### PR DESCRIPTION
**Pull request by ansiwen**
In order to avoid external dependencies on implementation details, this change replaces the error types with the unexported versions. In case some application really needs access to the integer value, it can use the pattern

```
var intErr interface{ Int() int }
if errors.As(err, intErr) { ... intErr.Int() ... }
```

Signed-off-by: Sven Anderson <sven@redhat.com>

### CODE REVIEW

**Comment by phlogistonjohn** [comment]
I generally think this is the right direction to go in. I've been somewhat annoyed by the three distinct, but mostly the same, error types the library has been using for some time. Making the types private and using interfaces is cleaner IMO.

One concern I do have is breaking current code. While we don't promise backwards compatibility until v1.0, I like to provide a smooth transition. One thing we could consider is to move all the internal code to the private type like you have but still provide a public type alias for one release. Then in the release notes, we would mention that the exported types (CephFSError, RadosError, RbdError) are deprecated are going to be removed in the next release.
Then a short pr just to remove the aliases can be done in the following cycle.

As for the implementation, what do you think about using a function name of `Errno` instead of `Int`? I don't think we ought to hide the fact that the api use Unix/Linux style errnos. Using Errno as the name reflects that and is more specific.

Just to double check none of these changes impact using the sentinel error style that many of the calling code does right? IOW this shouldn't effect stuff like `err == rbd.ErrFoo` or, with a newer style, `errors.Is(err, rbd.ErrFoo)`, IIUC.

**Comment by ansiwen** [comment]
> One concern I do have is breaking current code. While we don't promise backwards compatibility until v1.0, I like to provide a smooth transition. One thing we could consider is to move all the internal code to the private type like you have but still provide a public type alias for one release. Then in the release notes, we would mention that the exported types (CephFSError, RadosError, RbdError) are deprecated are going to be removed in the next release.
> Then a short pr just to remove the aliases can be done in the following cycle.

Sure, we can do that. I only checked ceph-csi as our "main" consumer and didn't find any references there. But if you are concerned, someone will have issues with this, we can go smoother. However, I think the effect will be the same, people will ignore it and then fix it with the next release. And it would be trivial fixes, so I would take the risk overall. But your call, I'm fine with both.

> As for the implementation, what do you think about using a function name of `Errno` instead of `Int`? I don't think we ought to hide the fact that the api use Unix/Linux style errnos. Using Errno as the name reflects that and is more specific.

Yeah, I like that. Errno() as an interface has more semantics.

> Just to double check none of these changes impact using the sentinel error style that many of the calling code does right? IOW this shouldn't effect stuff like `err == rbd.ErrFoo` or, with a newer style, `errors.Is(err, rbd.ErrFoo)`, IIUC.

Sure, I didn't change any existing code within the errors.go files, and the error values stay exported. However, I'm in favor of going a step further in another PR and changing the implementation so that the sentinels are never directly returned, but always in wrapped errors, so that you have to use errors.Is(), as recommended for package APIs [here](https://blog.golang.org/go1.13-errors#TOC_5.).

File: rados/errors.go [hunk 1]
```diff
@@ -12,27 +12,27 @@ import (
...
+// Error returns the error string for the radosError type.
+func (e radosError) Error() string {
 	errno, s := errutil.FormatErrno(int(e))
 	if s == "" {
 		return fmt.Sprintf("rados: ret=%d", errno)
 	}
 	return fmt.Sprintf("rados: ret=%d, %s", errno, s)
 }
 
+func (e radosError) Errno() int {
```
Comment (agarwal-mudit) [inline rados/errors.go:27]: This is newly added and not used anywhere. We can add a test case for this, may be in a different PR.

File: cephfs/errors.go [hunk 2]
```diff
@@ -11,41 +11,41 @@ import (
...
+// Error returns the error string for the cephFSError type.
+func (e cephFSError) Error() string {
 	errno, s := errutil.FormatErrno(int(e))
 	if s == "" {
 		return fmt.Sprintf("cephfs: ret=%d", errno)
 	}
 	return fmt.Sprintf("cephfs: ret=%d, %s", errno, s)
 }
 
+func (e cephFSError) Errno() int {
```
Comment (agarwal-mudit) [inline cephfs/errors.go:26]: For this one also.

**Comment by phlogistonjohn** [comment]
I'm merging this. @ansiwen could you file a follow up pr to add coverage for the rados and cephfs packages instances of the Errno function, as @agarwal-mudit  suggested? Thanks!
